lib/libzpool/zrepl_mgmt.c

```diff
@@ -180,6 +161,14 @@ uzfs_remove_zinfo_list(zvol_info_t *zinfo)
 	(void) pthread_mutex_unlock(&zinfo->zinfo_mutex);
 	/* Base refcount is droped here */
 	uzfs_zinfo_drop_refcnt(zinfo, B_TRUE);
+
+	/* Wait for refcounts to be drained */
+	while (zinfo->refcnt > 0) {
+		sleep(1);
```

Comment:
this is very rough algorithm and can wait forever if the operation which is currently in progress with zinfo takes very long time. The offline condition is checked only at certain places, i.e. when new command is read from connection. Now consider what happens if new command is never sent but the connection is not closed by iscsi target. We will be stuck here waiting for ever.

Reply:
We send signal to uzfs_zvol_io_ack_sender from uzfs_mark_offline_and_free_zinfo() during volume offline. uzfs_zvol_io_ack_sender, once wake up will observe volume state and close fd. By closing fd, uzfs_zvol_io_receiver thread will also come to know that fd got closed so it will take action and exit by droping refcount. I agree, that coming to know instantantly that "volume state has changed" is not possible(and that is not different from other storage product in market).
Behaviour of system was like that only even before this change. By adding this sleep/wait is to avoid issue DE37.

As per this explanation this operation wont wait forever based on above explanation.